File: backend/src/main/java/tom/projects/monkeylog/controller/ExerciseController.java
package tom.projects.monkeylog.controller;

import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import tom.projects.monkeylog.dto.workout.*;
import tom.projects.monkeylog.mapper.ExerciseMapper;
import tom.projects.monkeylog.model.exercise.ExerciseCategory;
import tom.projects.monkeylog.model.exercise.ExerciseType;
import tom.projects.monkeylog.service.ExerciseService;

import java.util.Arrays;
import java.util.List;
import java.util.UUID;

@RestController
@Tag(name = "Exercise")
@RequiredArgsConstructor
public class ExerciseController {
    private final ExerciseService exerciseService;
    private final ExerciseMapper exerciseMapper;

    @GetMapping("/exercises")
    List<ExerciseResponse> getExercises() {
        return exerciseMapper.exercisesToExerciseResponses(exerciseService.all());
    }

    @GetMapping("/exercises/{id}")
    ExerciseResponse getExercise(@PathVariable UUID id) {
        return exerciseMapper.exerciseToExerciseResponse(exerciseService.get(id));
    }

    @PostMapping("/exercises")
    ExerciseResponse createExercise(@RequestBody ExerciseCreateRequest exerciseCreateRequest) {
        return exerciseMapper.exerciseToExerciseResponse(exerciseService.save(exerciseCreateRequest));
    }

    @PatchMapping("/exercises/{id}")
    ExerciseResponse updateExercise(@RequestBody ExerciseUpdateRequest exerciseUpdateRequest, @PathVariable UUID id) {
        return exerciseMapper.exerciseToExerciseResponse(exerciseService.update(exerciseUpdateRequest, id));
    }

    @DeleteMapping("/exercises/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    void deleteExercise(@PathVariable UUID id) {
        exerciseService.delete(id);
    }

    @GetMapping("/exercises/types")
    List<ExerciseTypeResponse> getExerciseTypes() {
        return Arrays.stream(ExerciseType.values()).map(exerciseMapper::exerciseTypeToExerciseTypeResponse).toList();
    }

    @GetMapping("/exercises/categories")
    List<ExerciseCategoryResponse> getExerciseCategories() {
        return Arrays.stream(ExerciseCategory.values()).map(exerciseMapper::exerciseCategoryToExerciseCategoryResponse).toList();
    }
}

File: backendkt/src/main/kotlin/tom/com/monkeylog/controller/ExerciseController.kt
package tom.com.monkeylog.controller

import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import tom.com.monkeylog.dto.workout.ExerciseCreateRequest
import tom.com.monkeylog.dto.workout.ExerciseUpdateRequest
import tom.com.monkeylog.mapper.toResponse
import tom.com.monkeylog.model.exercise.ExerciseCategory
import tom.com.monkeylog.model.exercise.ExerciseType
import tom.com.monkeylog.service.ExerciseService
import java.util.*

@RestController
@Tag(name = "Exercise")
class ExerciseController(
    private val exerciseService: ExerciseService,
) {
    @GetMapping("/exercises")
    fun exercises() = exerciseService.allExercises().map { it.toResponse() }

    @GetMapping("/exercises/{id}")
    fun getExercise(@PathVariable id: UUID) = exerciseService.getExercise(id).toResponse()

    @PostMapping("/exercises")
    fun createExercise(@RequestBody exerciseCreateRequest: ExerciseCreateRequest) =
        exerciseService.saveExercise(exerciseCreateRequest).toResponse()

    @PatchMapping("/exercises/{id}")
    fun updateExercise(
        @RequestBody exerciseUpdateRequest: ExerciseUpdateRequest,
        @PathVariable id: UUID
    ) = exerciseService.updateExercise(exerciseUpdateRequest, id).toResponse()

    @DeleteMapping("/exercises/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteExercise(@PathVariable id: UUID) = exerciseService.deleteExercise(id)

    @GetMapping("/exercises/types")
    fun exerciseTypes() = ExerciseType.values().map { it.toResponse() }

    @GetMapping("/exercises/categories")
    fun exerciseCategories() = ExerciseCategory.values().map { it.toResponse() }
}
